include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1124 @@
+             const std::vector<scalar_type> &independent_variables)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+
+        double *f = new double();
+
+#ifdef DEAL_II_WITH_ADOLC
+        ::function(active_tape_index,
+                   1, // Only one dependent variable
+                   n_independent_variables,
+                   const_cast<double *>(independent_variables.data()),
+                   f);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        const double value = f[0];
+
+        // Cleanup :-/
+        delete f;
+        f = nullptr;
+
+        return value;
+      }
+
+      static void
+      gradient (const unsigned int             &active_tape_index,
+                const unsigned int             &n_independent_variables,
+                const std::vector<scalar_type> &independent_variables,
+                Vector<scalar_type>            &gradient)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(gradient.size() == n_independent_variables,
+               ExcMessage("The length of the gradient vector must be equal "
+                          "to the number of independent variables."));
+
+        scalar_type *g = new scalar_type[n_independent_variables]; // Use smart pointers or std_cxx11::array?
+
+#ifdef DEAL_II_WITH_ADOLC
+        ::gradient(active_tape_index,
+                   n_independent_variables,
+                   const_cast<scalar_type *>(independent_variables.data()),
+                   g);
+#else
+        AssertThrow(false, ExcNotImplemented());
+#endif
+
+        for (unsigned int i=0; i<n_independent_variables; ++i)
+          gradient[i] = g[i];
+
+        // Cleanup :-/
+        delete[] g;
+        g = nullptr;
+      }
+
+      static void
+      hessian (const unsigned int             &active_tape_index,
+               const unsigned int             &n_independent_variables,
+               const std::vector<scalar_type> &independent_variables,
+               FullMatrix<scalar_type>        &hessian)
+      {
+        Assert(independent_variables.size() == n_independent_variables,
+               ExcDimensionMismatch(independent_variables.size(),n_independent_variables));
+        Assert(hessian.m() == n_independent_variables,
+               ExcMessage("The (symmetric) hessian row length must be equal "
+                          "to the number of independent variables."));
+        Assert(hessian.n() == n_independent_variables,
+               ExcMessage("The (symmetric) hessian column length must be equal "
+                          "to the number of independent variables."));
+
+        scalar_type **H = new scalar_type*[n_independent_variables]; // Use smart pointers or std_cxx11::array?
```

Comment:
why not use `hessian` directly? alignment is exactly the same.

Reply:
Two reasons: 
1. The last parameter to `::hessian` must be a `double**`. There will be no way to safely extract an equivalent data type from `FullMatrix`.
2. `scalar_type **H` is not actually a full matrix, but only the upper triangle of the matrix. Adol-C takes advantage of the symmetry of the 2nd derivative operation.

Reply:
(1) same as above, have you tried assigning addresses of each rows of `FullMatrix`? (2) that's not an issue, you just need to copy the part from upper triangle before returning

Reply:
more precisely, give the address of the first element in each row of FullMatrix